text/0000-make-cargo-install-extensible.md

```diff
@@ -0,0 +1,122 @@
+- When re-installing a binary package this might run a different `install.rs` than the one that was initially run. This might lead to inconsistent/ undefined behavior.
+
+More corner cases would appear if `cargo install` was ever extended beyond binary applications. The handling of those corner cases is left as an exercise to the writers of a proposal to extend `cargo install` to libraries or other types of files/applications.
+
+## Drawbacks
+
+- This adds another mechanism to run code from the crate on the user’s computer. However, `build.rs` already supports that, so building a crate already involves running code from that crate. Increasing the surface area of the crate code being run is potentially more concerning. 
+## Rationale and alternatives
+
+- There is already a way for custom code to be run during a build step via the `build.rs` file which is run before the build process of a crate. This proposal adds analogous functionality to the install process of a crate.
+  - Tasks defined in `install.rs` can depend on a completed build or the specific platform upon which the application is being installed (e.g. paths), while `build.rs` is run before the main compilation step.
+- Not allowing for the extension of `cargo install` relegates installation steps to the end user and any sort of post-install documentation. This has several downsides:
+  - Forces users to follow additional install documentation when installing tools.
+  - Makes the packaging of applications more complicated for developers of tools which can have negative impact on the ecosystem as a whole.
+    - Tools may be written in different languages/ toolkits.
+    - Developers are forced to implement their own installation mechanisms.
+    - Tools only support specific popular platforms because being cross-platform is deemed “too difficult”.
+- Making `cargo install` more capable could encourage people to use it as a primary distribution mechanism for a broader class of applications, rather than just for simple command-line tools. On the other hand, this same mechanism can also serve as the basis for distribution packaging, which typically wants to install into a temporary directory and package the result.
+- It makes the emergence of conventions significantly more difficult as the option to reuse, share, or automate this task has significantly affordance than integrating it into the familiar `cargo` step.
+- Currently, `cargo install` tracks what files it installs, and supports `cargo uninstall`; this extension mechanism does not hook into that. Potentially, `install.rs` could emit a list of installed files and let `cargo install` install them, producing a log of those files for later uninstallation. However, we do not intend for `cargo install` to become a full-featured package management mechanism; rather, we expect `cargo install` to work analogously to `make install`. While the occasional package provides a `make uninstall`, few developers expect such a mechanism.
```

Comment:
Do we want to emulate `make install` tho? I've never found `make` to be a reliable program.
I also question the assertion on expectations. What is the evidence for it?

Reply:
I think this is fair. The relation to `make` was in a convo with @joshtriplett @alexcrichton and @withoutboats - I also share your concern a bit re this, but I found that the affordances this gives us for both `cargo install` distribution AS WELL as platform specific package managers outweighed it in my opinion. I'm not sure if there's any data/anecdata to be found but I would be happy to try to find some!

Reply:
If we did want anecdata, I wonder if cross compilers and distro packagers would be a good source of information on `make install` behaviour as they're groups of people likely to spend a high percentage of their time compiling things.

My own experience/expectation from when I was doing a lot of cross-compilation with emscripten is that things supporting `make install` a) permit some way to set the installation 'prefix' (be it `--prefix` with autotools and autotools-mimicking things like the rust configure script or the cmake variable or something else) and nothing gets installed outside of that prefix (e.g. completions end up in `$PREFIX/etc/bash_completion.d)` and b) `make uninstall` is a bit flaky.

Reply:
Small note: I don't think cross compiler and distro packagers are representative of the set "developers". ;) Even so, any and all anecdata would be nice to have even if unrepresentative.

Reply:
Ah I take your point, perception is important. I was just thinking about the most efficient way to get feedback on large numbers of executed `make install` commands.